pkg/vault/contracts/Vault.sol

```diff
@@ -376,6 +372,8 @@ contract Vault is IVaultMain, VaultCommon, Proxy {
             poolSwapParams.amountGivenScaled18 -= locals.totalSwapFeeAmountScaled18;
         }
 
+        _ensureValidSwapAmount(poolSwapParams.amountGivenScaled18);
```

Comment:
Should we apply this check before the tax? 

The dynamic swap fee of 100% is ill-defined anyways; you can execute it exact-in like this, but exact-out will end up dividing by 0 [here](https://github.com/balancer/balancer-v3-monorepo/pull/991/files#diff-f03125e354a92ba020d9db1ee551c1b9ba5034e99ae369a139e02e8679002709R409).

Reply:
I had it that way in the beginning - but we're doing amountOut after adjustment, so I'm not sure. The "after tax" amount is what's going into the pool math (i.e., a high fee could take a small amount given and make it near zero, which we're trying to avoid).

Similarly, maybe we should do amountOut before adjustment (right after the pool math). So we'd just bracket the pool math:

_ensureValidSwapAmount(poolSwapParams.amountGivenScaled18);

amountCalculatedScaled18 = IBasePool(vaultSwapParams.pool).onSwap(poolSwapParams);

_ensureValidSwapAmount(amountCalculatedScaled18);